Close #LGVISIUM-60: Elevation extraction

dcleres opened the pull request:
This PR adds the functionality to extract the elevation from the borehole profiles. The extracted information is then saved into the metadata part of the document. 

The current metrics on the Zurich dataset of this feature are: 
- 'elevation_accuracy': 0.68125,
- 'elevation_precision': 0.9008264462809917,
- 'elevation_recall': 0.7364864864864865,
- 'elevation_f1': 0.8104089219330854,
- 'elevation_tp': 109,
- 'elevation_fp': 12,
- 'elevation_fn': 39

Potential improvements for future issues: 
- Use the elevation information to compute the depth of the groundwater 
Currently, some functions and methods of the groundwater extraction and elevation extraction are quite redundant. It would make sense to refactor the code to have a parent class that implements some of the logic. 
- Currently, I did not spend a lot of time trying to optimize the performance of the extraction. There are certainly still a few easy wins to achieve here by taking more time to work on this feature extractor.

dcleres commented:
Latest performance: 'elevation_accuracy': 0.7028571428571428, 'elevation_precision': 0.8913043478260869, 'elevation_recall': 0.76875, 'elevation_f1': 0.825503355704698, 'elevation_tp': 123, 'elevation_fp': 15, 'elevation_fn': 37,

stijnvermeeren-swisstopo commented on src/stratigraphy/elevation/elevation_extraction.py:
```diff
@@ -0,0 +1,268 @@
+        }
+
+
+class ElevationExtractor:
```
There is a LOT of duplicated code here with the groundwater extractor. What stops us from creating a parent class with the shared logic, and instantiate GroundwaterExtractor and ElevationExtractor as child classes that only need to implement their respective unique functionality?

dcleres replied:
Nothing stop us! :-) 

I was mainly trying to see how we merge the Groundwater level extractor and I will refactor this today. Thank you for the feedback!

stijnvermeeren-swisstopo commented on src/stratigraphy/elevation/elevation_extraction.py:
```diff
@@ -0,0 +1,268 @@
+                if match:
+                    matches.append((line, sum(match.fuzzy_counts)))
+
+        # if no match was found, return None
+        if len(matches) == 0:
+            return None
+
+        # Remove duplicates
+        matches = list(dict.fromkeys(matches))
+
+        # Sort the matches by their error counts (ascending order)
+        matches.sort(key=lambda x: x[1])
+
+        # Return the top three matches (lines only)
+        return [match[0] for match in matches[:5]]
```
Can we remove the selection of the "best n matched" and improve the deduplication, analogous to https://github.com/swisstopo/swissgeol-boreholes-dataextraction/pull/72/commits/ee46a8267fdff082660ad5f7a0a107fe70e51ea9 ?

dcleres replied:
Yes I will do so.

stijnvermeeren-swisstopo reviewed (commented):
Why did it pick the number "682.783" instead of "408.40" in this document? (268124094-bp.pdf)

![image](https://github.com/user-attachments/assets/4387638b-e2ae-45dd-b131-5c8825158be4)

dcleres commented:
> Why did it pick the number "682.783" instead of "408.40" in this document? (268124094-bp.pdf)
> 
> ![image](https://private-user-images.githubusercontent.com/144008419/363146586-4387638b-e2ae-45dd-b131-5c8825158be4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUwMjAyNjYsIm5iZiI6MTcyNTAxOTk2NiwicGF0aCI6Ii8xNDQwMDg0MTkvMzYzMTQ2NTg2LTQzODc2MzhiLWUyYWUtNDVkZC1iMTMxLTVjODgyNTE1OGJlNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgzMFQxMjEyNDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xOTY3NTBmNjIzNDk5NjE1ZDcyYzkzMTM4MThiMjY2MWE1Nzc2OGM3ZjgxMTg4NjY1YzA1YjhmNDg3ZjQ5MGQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.rNiFNyGXmmmHnoaO-EruBqj4IJqEshvG0IdYTaQQ2Jw)
> 
> Edit: nevermind, the "408.40" seems to be missing from the OCR, so that's not something that we can deal with in the context of elevation extraction...

There are indeed a few other examples for this behavior. I tried to list them in https://jira.swisstopo.ch/projects/LGVISIUM/issues/LGVISIUM-62?filter=allopenissues.

stijnvermeeren-swisstopo commented on src/stratigraphy/data_extractor/data_extractor.py:
```diff
@@ -0,0 +1,143 @@
+                if match:
+                    matches.append((line, sum(match.fuzzy_counts)))
+
+        # if no match was found, return None
+        if len(matches) == 0:
+            return None
+
+        # Remove duplicates
+        matches = list(dict.fromkeys(matches))
```
For groundwater extraction, I implemented this differently in https://github.com/swisstopo/swissgeol-boreholes-dataextraction/pull/72/commits/ee46a8267fdff082660ad5f7a0a107fe70e51ea9 . I believe that that that implementation might also be better for the other attributes.

Maybe this is also what is causing the merge conflict that Github is currently reporting?

dcleres replied:
Thank you for the detailed look into the PR. Yes, I do believe that I still need to merge with the latest main branch which will add your changes. Currently, there is also still the n-best logic in there that we removed.

dcleres commented:
Thank you for the detailed look into the PR. Yes, I do believe that I still need to merge with the latest main branch which will add your changes. Currently, there is also still the n-best logic in there that we removed. On 2 Sep 2024, at 17:01, Stijn Vermeeren ***@***.***> wrote:﻿
@stijnvermeeren-swisstopo requested changes on this pull request.



In src/stratigraphy/data_extractor/data_extractor.py:
> +    """Abstract class for data extraction from stratigraphy data files.
+
+    This class defines the interface for extracting data from stratigraphy data files. Subclasses must implement the
+    extract_data method to define the data extraction logic.
+    """
+
+    doc: fitz.Document = None
+    feature_keys: list[str] = None
+    feature_name = None
+
+    def __init__(self, document: fitz.Document, feature_name: str = "coordinate"):
+        """Initializes the CoordinateExtractor object.
+
+        Args:
+            document (fitz.Document): A PDF document.
+            feature_name (str, optional): The name of the feature to extract. Defaults to "coordinate".

why have a default value at all for this?



In src/stratigraphy/data_extractor/data_extractor.py:
> +logger = logging.getLogger(__name__)
+
+
+class DataExtractor(ABC):
+    """Abstract class for data extraction from stratigraphy data files.
+
+    This class defines the interface for extracting data from stratigraphy data files. Subclasses must implement the
+    extract_data method to define the data extraction logic.
+    """
+
+    doc: fitz.Document = None
+    feature_keys: list[str] = None
+    feature_name = None
+
+    def __init__(self, document: fitz.Document, feature_name: str = "coordinate"):
+        """Initializes the CoordinateExtractor object.

Make the description more generic



In src/stratigraphy/data_extractor/data_extractor.py:
> +        """Finds the location of a coordinate key in a string of text.
+
+        This is useful to reduce the text within which the coordinates are searched. If the text is too large
+        false positive (found coordinates that are no coordinates) are more likely.

Make the description more generic.



In src/stratigraphy/main.py:
> @@ -172,24 +172,33 @@ def start_pipeline(
                     predictions[filename]["language"] = language
 
                     # Extract the coordinates of the borehole
-                    coordinate_extractor = CoordinateExtractor(doc)
-                    coordinates = coordinate_extractor.extract_coordinates()
+                    coordinate_extractor = CoordinateExtractor(
+                        document=doc,
+                        feature_name="coordinate",

Wouldn't it be better to fix the feature_name in the CoordinateExtractor class definition? Because we will never instantiate a CoordinateExtractor with a different feature_name, or will we?
(Same for elevation and groundwater.)



In src/stratigraphy/data_extractor/data_extractor.py:
> +            if self.feature_name in ["elevation", "coordinate"]:
+                if found_feature_value:
+                    feature_value = getattr(found_feature_value, self.feature_name)
+                    logger.info(f"Found {self.feature_name} information on page {page_number}: {feature_value}")
+                    return found_feature_value
+            elif self.feature_name == "groundwater":
+                if len(found_feature_value):
+                    groundwater_output = ", ".join(
+                        [str(getattr(entry, self.feature_name)) for entry in found_feature_value]
+                    )
+                    logger.info(f"Found groundwater information on page {page_number}: {groundwater_output}")
+                    return found_feature_value
+            else:
+                raise ValueError(f"Feature name '{self.feature_name}' not supported")

It might be cleaner to implement the in this respected child classes, instead of having this if-else logic here?



In src/stratigraphy/data_extractor/data_extractor.py:
> +        """Extracts the groundwater information from a borehole profile.
+
+        Processes the borehole profile page by page and tries to find the coordinates in the respective text of the
+        page.
+        Algorithm description:
+            1. if that gives no results, search for coordinates close to an explicit "groundwater" label (e.g. "Gswp")
+
+        Returns:
+            list[GroundwaterInformationOnPage] | None: the extracted coordinates (if any)

generalise the description and fix the return type



In src/stratigraphy/data_extractor/data_extractor.py:
> +            if self.feature_name in ["elevation", "coordinate"]:
+                if found_feature_value:
+                    feature_value = getattr(found_feature_value, self.feature_name)
+                    logger.info(f"Found {self.feature_name} information on page {page_number}: {feature_value}")
+                    return found_feature_value
+            elif self.feature_name == "groundwater":
+                if len(found_feature_value):
+                    groundwater_output = ", ".join(
+                        [str(getattr(entry, self.feature_name)) for entry in found_feature_value]
+                    )
+                    logger.info(f"Found groundwater information on page {page_number}: {groundwater_output}")
+                    return found_feature_value
+            else:
+                raise ValueError(f"Feature name '{self.feature_name}' not supported")
+
+        logger.info("No groundwater found in this borehole profile.")

generalise the log message



In src/stratigraphy/data_extractor/data_extractor.py:
> +        if self.feature_name == "coordinate":
+            # Return the best match (line only)
+            best_match = min(matches, key=lambda x: x[1])
+            return best_match[0]
+        elif self.feature_name in ["elevation", "groundwater"]:
+            # Sort the matches by their error counts (ascending order)
+            matches.sort(key=lambda x: x[1])
+
+            # Return the top three matches (lines only)
+            return [match[0] for match in matches[:5]]
+        else:
+            raise ValueError(f"Feature name '{self.feature_name}' not supported.")

It might be cleaner to implement the in this respected child classes, instead of having this if-else logic here?



In src/stratigraphy/data_extractor/data_extractor.py:
> +            else:
+                pattern = regex.compile(r"\b(" + key + "){e<" + str(allowed_errors) + r"}\b", flags=regex.IGNORECASE)
+
+            for line in lines:
+                match = pattern.search(line.text)
+                if match:
+                    matches.append((line, sum(match.fuzzy_counts)))
+
+        # if no match was found, return None
+        if len(matches) == 0:
+            return None
+
+        # Remove duplicates
+        matches = list(dict.fromkeys(matches))
+
+        # OPTION 1

what's "option 1"?



In src/stratigraphy/data_extractor/data_extractor.py:
> +        # if no match was found, return None
+        if len(matches) == 0:
+            return None
+
+        # Remove duplicates
+        matches = list(dict.fromkeys(matches))

For groundwater extraction, I implemented this differently in ee46a82 . I believe that that that implementation might also be better for the other attributes.
Maybe this is also what is causing the merge conflict that Github is currently reporting?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were assigned.Message ID: ***@***.***>

dcleres commented on src/stratigraphy/main.py:
```diff
@@ -172,24 +172,33 @@ def start_pipeline(
-                    coordinates = coordinate_extractor.extract_coordinates()
+                    coordinate_extractor = CoordinateExtractor(
+                        document=doc,
+                        feature_name="coordinate",
```
Indeed, good point.

stijnvermeeren-swisstopo commented on src/stratigraphy/groundwater/utility.py:
```diff
@@ -64,16 +64,20 @@ def extract_elevation(text: str) -> float | None:
+        r"(\d+(\.\d+)?)\s*m\s*ur.",
+        r"(\d{3,}\.\d+)",
+        r"(\d{3,})\s*m",
+        # r"([\d]+(?:\s*\.\s*\d+))",  # match numbers that may have a space before and after the decimal point.
```
Code commented out. Can we delete this?

stijnvermeeren-swisstopo commented on src/stratigraphy/elevation/elevation_extraction.py:
```diff
@@ -0,0 +1,189 @@
+                # or XXXX # Add other techniques here
+            )
+
+            if found_feature_value:
+                feature_value = getattr(found_feature_value, self.feature_name)
+                logger.info("Found %s on page %s: %s", self.feature_name, page_number, feature_value)
+                return found_feature_value
+
+        logger.info("No %s found in this borehole profile.", self.feature_name)
```
this code no longer needs to be so generic, now that this method is back in the ElevationExtractor class, and no longer in the abstract parent class.

dcleres commented on config/matching_params.yml:
```diff
@@ -154,3 +154,15 @@ groundwater_unit_keys:
+  - OK Terrain
+  - OKT
+  - Ansatzhöhe
+  # - OKR
```
It was generating false postives. Usually, OKR was associated with the elevation of the drill piece. Therefore, I preferred to remove it. But I will delete it.
